File: src/examples/USStreetLookupsWithMatchStrategyExamples.cs
namespace Examples
{
	using System;
	using System.Collections.Generic;
	using System.IO;
	using SmartyStreets;
	using SmartyStreets.USStreetApi;

	internal static class USStreetLookupsWithMatchStrategyExamples
	{
		public static void Run()
		{
			// You don't have to store your keys in environment variables, but we recommend it.
			var authId = Environment.GetEnvironmentVariable("SMARTY_AUTH_ID");
			var authToken = Environment.GetEnvironmentVariable("SMARTY_AUTH_TOKEN");

			using var client = new ClientBuilder(new BasicAuthCredentials(authId, authToken)).BuildUsStreetApiClient();
			var batch = new Batch();

			// Each address is run through all three match strategies so you can compare how
			// 'strict', 'enhanced', and 'invalid' each handle a valid, an invalid, and an
			// ambiguous address.
			//   - strict:   only returns candidates that are valid, mailable addresses.
			//   - enhanced: returns a more comprehensive dataset (requires a US Core or Rooftop license).
			//   - invalid:  most permissive; always returns at least one candidate (a best-guess standardization).
			// Documentation for input fields: https://smartystreets.com/docs/us-street-api#input-fields
			var addresses = new[]
			{
				new { Label = "valid (real, deliverable)",    Street = "1600 Amphitheatre Pkwy", City = "Mountain View", State = "CA", Zip = "94043" },
				new { Label = "invalid (no such address)",    Street = "9999 W 1150 S",          City = "Provo",         State = "UT", Zip = "84601" },
				new { Label = "ambiguous (missing ZIP/unit)", Street = "1 Rosedale St",          City = "Baltimore",     State = "MD", Zip = "" },
			};
			var strategies = new[] { Lookup.STRICT, Lookup.ENHANCED, Lookup.INVALID };

			// parallel metadata for each lookup, in the order they are added to the batch
			var cases = new List<(string Label, string Address, string Strategy)>();

			try
			{
				foreach (var address in addresses)
				{
					foreach (var strategy in strategies)
					{
						batch.Add(new Lookup
						{
							Street = address.Street,
							City = address.City,
							State = address.State,
							ZipCode = address.Zip,
							MatchStrategy = strategy,
							MaxCandidates = 10 // allow ambiguous addresses to return more than one match
						});
						cases.Add((address.Label, $"{address.Street}, {address.City}, {address.State}", strategy));
					}
				}

				client.Send(batch);
			}
			catch (BatchFullException)
			{
				Console.WriteLine("Error. The batch is already full.");
				return;
			}
			catch (SmartyException ex)
			{
				Console.WriteLine(ex.Message);
				Console.WriteLine(ex.StackTrace);
				return;
			}
			catch (IOException ex)
			{
				Console.WriteLine(ex.StackTrace);
				return;
			}

			string lastAddress = null;
			for (var i = 0; i < batch.Count; i++)
			{
				var (label, address, strategy) = cases[i];

				if (address != lastAddress)
				{
					Console.WriteLine("\n" + new string('=', 70));
					Console.WriteLine($" Address: {address}  [{label}]");
					Console.WriteLine(new string('=', 70));
					lastAddress = address;
				}

				var candidates = batch[i].Result;
				Console.WriteLine($"\n--- '{strategy}' strategy ---");

				if (candidates.Count == 0)
				{
					Console.WriteLine("  0 candidates - no match returned under this strategy.");
					continue;
				}

				Console.WriteLine($"  {candidates.Count} candidate(s):");
				foreach (var candidate in candidates)
					Console.WriteLine($"    [{candidate.CandidateIndex}] {candidate.DeliveryLine1}  {candidate.LastLine}");
			}
		}
	}
}

File: src/vbexamples/USStreetLookupsWithMatchStrategyExamples.vb
Imports System.Data
Imports System.IO
Imports System.Text.RegularExpressions
Imports SmartyStreets
Imports SmartyStreets.USStreetApi

Module USStreetLookupsWithMatchStrategyExamples

    Dim authID = Environment.GetEnvironmentVariable("SMARTY_AUTH_ID")
    Dim authToken = Environment.GetEnvironmentVariable("SMARTY_AUTH_TOKEN")
    Dim url = Environment.GetEnvironmentVariable("SMARTY_US_STREET_URL")

    Dim client = New ClientBuilder(authID, authToken).WithLicense(New List(Of String) From {"us-core-cloud"}).WithCustomBaseUrl(url).BuildUsStreetApiClient()

    Sub USStreetLookupsWithMatchStrategyExamples()

        Dim batch = New Batch()

        Dim addressWithStrictStrategy As New Lookup()
        With addressWithStrictStrategy
            .InputId = "23493"
            .Street = "691 W 1150 S"
            .City = "provo"
            .State = "utah"
            .MatchStrategy = Lookup.STRICT
        End With

        Dim addressWithRangeStrategy As New Lookup()
        With addressWithRangeStrategy
            .Street = "691 W 1150 S"
            .City = "provo"
            .State = "utah"
            .MatchStrategy = Lookup.ENHANCED
        End With

        Dim addressWithInvalidStrategy As New Lookup()
        With addressWithInvalidStrategy
            .Street = "691 W 1150 S"
            .City = "provo"
            .State = "utah"
            .MatchStrategy = Lookup.INVALID
        End With

        Console.WriteLine("*******************************************************")
        Console.WriteLine()

        Try
            batch.Add(addressWithStrictStrategy)
            batch.Add(addressWithRangeStrategy)
            batch.Add(addressWithInvalidStrategy)
            client.Send(batch)
        Catch ex As BatchFullException
            Console.WriteLine("Error. The batch is already full.")
            Console.WriteLine(ex.Message)
            Console.WriteLine(ex.StackTrace)
        Catch ex As IOException
            Console.WriteLine(ex.StackTrace)
        Catch ex As Exception
            Console.WriteLine(ex.Message)
            Console.WriteLine(ex.StackTrace)
        End Try

        Dim numLookups = batch.Count

        For i As Integer = 0 To numLookups - 1
            If i > 0 Then
                Console.WriteLine()
            End If

            Dim candidates = batch(i).Result
            Console.WriteLine("Address " + CStr(i) + " original lookup: " + batch(i).Street + ", " + batch(i).Street2 + ", " + batch(i).Secondary + ", " + batch(i).City + ", " + batch(i).State + ", " + batch(i).ZipCode + Environment.NewLine())
            If candidates.Count = 0 Then
                Console.WriteLine("Address " + CStr(i) + " has no candidates. The address is not valid." + Environment.NewLine)
                Continue For
            End If

            Console.WriteLine("Address " + CStr(i) + " has " + CStr(candidates.Count) + " candidate" + If(candidates.Count = 1, "", "s") + Environment.NewLine + "If the match parameter is set to STRICT, the address is valid." + Environment.NewLine + "Otherwise, check the Analysis output fields to see if the address is valid." + Environment.NewLine())

            Console.WriteLine("Input ID: " + batch(i).InputId)

            For Each candidate In candidates
                Console.WriteLine()
                Dim components = candidate.Components
                Dim metadata = candidate.Metadata

                Console.Write("Candidate " + CStr(candidate.CandidateIndex))
                Dim match = batch(i).MatchStrategy
                If match Is Nothing Then
                    match = "strict"
                End If

                Console.WriteLine(" with " + match + " strategy:")
                Console.WriteLine("Delivery line 1: " + candidate.DeliveryLine1)
                Console.WriteLine("Last line:       " + candidate.LastLine)
                Console.WriteLine("ZIP Code:        " + components.ZipCode + "-" + components.Plus4Code)
                Console.WriteLine("County:          " + metadata.CountyName)
                Console.WriteLine("Latitude:        " + CStr(metadata.Latitude))
                Console.WriteLine("Longitude:       " + CStr(metadata.Longitude))
            Next

        Next

        Console.WriteLine()

    End Sub

End Module
